extras/flake-typescript.nix

```diff
@@ -0,0 +1,186 @@
+        # dependencies from `nix`.
+        mkdir .nix-node-deps/
+
+        ${ if builtins.length dependencies == 0 then "" else
+            ''
+                ${builtins.concatStringsSep "\n" (builtins.map (pkgPath: 
+                    ''
+                        echo "Copying ${pkgPath}..."
+                        cp "${pkgPath}" .nix-node-deps/
+                    ''
+                        )dependencies)}
+
+                echo 'Running `npm install`...'
+                HOME=$TMPDIR npm install --save --package-lock-only ${builtins.concatStringsSep " " (builtins.map (pkgPath: ''".nix-node-deps/$(basename "${pkgPath}")"'') dependencies) }
+            ''
+        }
+
+        #########################################################
+        # Run `node2nix`
+        #########################################################
+        echo 'Running `node2nix`...'
+        node2nix --input package.json --lock package-lock.json ${builtins.concatStringsSep " " extraFlags}
+
+        # Reset the permissions for  `package.json` and `package-lock.json` to
+        # read only for everyone.
+        chmod =444 package.json
+        chmod =444 package-lock.json
+      '';
+
+  node2nixDevelop = node2nixExprs { extraFlags = [ "--development" ]; };
+  node2nixDevelopAttrs = ((import node2nixDevelop) { inherit nodejs pkgs system; });
+
+  # Build the project (runs `npm run build`), and puts the entire output in the
+  # nix store
+  project = node2nixDevelopAttrs.package.override
+    {
+      postInstall =
+        ''
+          npm run --loglevel=verbose build
+        '';
+    };
+
+  shell = node2nixDevelopAttrs.shell.override
+    { };
+
+  # Creates a tarball of `project` using `npm pack` and puts it in the nix
+  # store.
+  tarball = pkgs.stdenv.mkDerivation {
+    buildInputs = [ nodejs ];
+    name = "${node2nixDevelopAttrs.args.name}.tgz";
+
+    # A glance at the generated nix expressions in `node2nixDevelop` will
+    # show why the following path is where the package really is.
+    src = "${project}/lib/node_modules/${node2nixDevelopAttrs.args.packageName}";
+    buildPhase = ''
+      cp -r "$src" .
+
+      # Why do we set `HOME=$TMPDIR`? This is because apparently `npm` will
+      # attempt to do something like `mkdir $HOME` for which `$HOME` is not
+      # writable.. so we fix this by making `$HOME` a writable directory.
+      tgzFile=$(HOME=$TMPDIR npm pack | tail -n 1)
+    '';
+    installPhase = ''
+      mv "$tgzFile" "$out"
+    '';
+
+  };
+
+  # Run tests with `npm test`.
+  test = node2nixDevelopAttrs.package.override
+    {
+      postInstall =
+        ''
+          npm --loglevel=verbose test
+          rm -rf $out
+          touch $out
+        '';
+    };
+
+in
+{
+  devShells = {
+    # Note: when using the devshell, when in the project root directory, type
+    # ```
+    # ln -s "$NODE_PATH" node_modules
+    # ```
+    # so when `node` executes, it will use the dependencies gathered together
+    # by nix
```

Comment:
Why? Why don't you just link these? Also this won't work right if NODE_PATH contains more than one paths separated by a :